File: 46-全排列.java
// 给定一个 没有重复 数字的序列，返回其所有可能的全排列。

// 示例:
// 输入: [1,2,3]
// 输出:
// [
//   [1,2,3],
//   [1,3,2],
//   [2,1,3],
//   [2,3,1],
//   [3,1,2],
//   [3,2,1]
// ]

import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;
import java.util.stream.Collectors;

class Solution {
    private final List<List<Integer>> res = new LinkedList<>();

    public List<List<Integer>> permute(int[] nums) {
        backtrace(Arrays.stream(nums).boxed().collect(Collectors.toList()), new LinkedList<>());
        return res;
    }

    private void backtrace(List<Integer> supplier, List<Integer> one) {
        if (supplier.isEmpty()) { // end of dfs
            res.add(one);
            return;
        }

        for (int i = 0; i < supplier.size(); i++) {
            var li = new LinkedList<>(supplier); // replication
            var newOne = new LinkedList<>(one); // replication
            var elem = li.remove(i);
            newOne.add(elem);
            backtrace(li, newOne);
        }
    }
}

File: 46. 全排列.py
# 给定一个不含重复数字的数组 nums ，返回其 所有可能的全排列 。你可以 按任意顺序 返回答案。
#
# 示例 1：
# 输入：nums = [1,2,3]
# 输出：[[1,2,3],[1,3,2],[2,1,3],[2,3,1],[3,1,2],[3,2,1]]
#
# 示例 2：
# 输入：nums = [0,1]
# 输出：[[0,1],[1,0]]
#
# 示例 3：
# 输入：nums = [1]
# 输出：[[1]]
#
# 提示：
# 1 <= nums.length <= 6
# -10 <= nums[i] <= 10
# nums 中的所有整数 互不相同

class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        self.visited = [False for i in range(len(nums))]
        self.ans = []

        def dfs(li):
            if len(li) == len(nums):
                self.ans.append(li[:])
                return
            for i, v in enumerate(self.visited):
                if v:
                    continue
                self.visited[i] = True
                li.append(nums[i])
                dfs(li)
                li.pop()
                self.visited[i] = False

        dfs([])
        return self.ans
